File: backend/parser/src/main/java/com/epam/deltix/quantgrid/parser/ast/BinaryOperation.java
package com.epam.deltix.quantgrid.parser.ast;

import lombok.AllArgsConstructor;
import lombok.Getter;

import java.util.Arrays;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@Getter
@AllArgsConstructor
public enum BinaryOperation {
    ADD("+", false),
    SUB("-", false),
    MUL("*", false),
    DIV("/", false),
    POW("^", false),
    LT("<", true),
    GT(">", true),
    LTE("<=", true),
    GTE(">=", true),
    NEQ("<>", true),
    EQ("=", true),
    AND("AND", false),
    OR("OR", false),
    MOD("MOD", false),
    CONCAT("&", true);

    private static final Map<String, BinaryOperation> OPERATIONS = Arrays.stream(BinaryOperation.values())
            .collect(Collectors.toUnmodifiableMap(BinaryOperation::getSymbol, Function.identity()));

    private final String symbol;
    private final boolean allowStrings;

    public static BinaryOperation parse(String s) {
        BinaryOperation operation = OPERATIONS.get(s);
        if (operation == null) {
            throw new IllegalArgumentException(s);
        }

        return operation;
    }
}

File: frontend/libs/parser/src/lib/ast/BinaryOperation.ts
export enum BinaryOperation {
  ADD = '+',
  SUB = '-',
  MUL = '*',
  DIV = '/',
  POW = '^',
  LT = '<',
  GT = '>',
  LTE = '<=',
  GTE = '>=',
  NEQ = '<>',
  EQ = '=',
  AND = 'AND',
  OR = 'OR',
  MOD = 'MOD',
}

export function parseBinaryOperation(s: string): BinaryOperation {
  switch (s) {
    case '+': {
      return BinaryOperation.ADD;
    }
    case '-': {
      return BinaryOperation.SUB;
    }
    case '*': {
      return BinaryOperation.MUL;
    }
    case '/': {
      return BinaryOperation.DIV;
    }
    case '^': {
      return BinaryOperation.POW;
    }
    case '<': {
      return BinaryOperation.LT;
    }
    case '>': {
      return BinaryOperation.GT;
    }
    case '<=': {
      return BinaryOperation.LTE;
    }
    case '>=': {
      return BinaryOperation.GTE;
    }
    case '<>': {
      return BinaryOperation.NEQ;
    }
    case '=': {
      return BinaryOperation.EQ;
    }
    case 'AND': {
      return BinaryOperation.AND;
    }
    case 'OR': {
      return BinaryOperation.OR;
    }
    case 'MOD': {
      return BinaryOperation.MOD;
    }
    default: {
      throw new Error('IllegalStateException');
    }
  }
}
